src/commands/buffer.rs

```diff
@@ -126,7 +126,11 @@ impl Buffer {
             bail!("Invalid size for buffer: {}", size);
         }
 
+        let mem_size = self.data_buffer.capacity();
         self.data_buffer.resize(size, 0);
+        if mem_size > 8192 && size < mem_size {
```

Comment:
One other thought: I think we should add another threshold to only shrink the buffer if we can reclaim at least a certain amount of memory by doing so. Reallocating the buffer is a relatively expensive operation, and it's not worth doing to save just a few bytes.

We should also uses consts to define these limits:

```
const BUFFER_RECLAIM_THRESHOLD: usize = 8192; // reclaim memory if buffer exceeds this size
const BUFFER_MIN_RECLAIM_BYTES: usize = 1024; // only reclaim if we can free at least this much memory
```

```suggestion
        if mem_size > BUFFER_RECLAIM_THRESHOLD && mem_size - size > BUFFER_MIN_RECLAIM_BYTES {
```

Reply:
I'll make it a constant. I kind of assumed you'd change the code with your preferred naming etc anyway, so I didn't bother. I don't mind doing it though, so I'll update later.

I am not really sold on the reclaim threshold, because at the end of the day this has the same problem. What if you can reclaim threshold+8bytes. Might as well have set the threshold higher in the first place. I think there this a good reason to make it a config setting on the client. You should know what the typical 99%ile of your payload size is and be able to size buffers accordingly. I can see if I can make that change too if you'd prefer.

Reply:
Looking at the code I could make it property of both the Buffer as well as ClientPolicy. That way it would be configurable and it would require relatively few changes.

Reply:
IMO we should make these settings public and dynamic so that users can tweak them, and then set intelligent defaults for them. Making them `const` will guarantee that the client will perform significantly worse for some users.

Reply:
I agree that `BUFFER_RECLAIM_THRESHOLD` should be configurable and the ClientPolicy is the right place for that.

Also, if you have a fairly predictable, and mostly constant payload size, then `BUFFER_RECLAIM_THRESHOLD` is indeed sufficient, because you can just set it slightly higher than your expected max. payload size. I think the `BUFFER_MIN_RECLAIM_BYTES` would be useful if that's not the case, i.e. if your payload size varies a lot. In that case the min. reclaim bytes limit would reduce the number of buffer reallocations the client has to do. It's a trade-off between memory usage and performance.